examples/notebooks/balancing-power.ipynb

```diff
@@ -0,0 +1,477 @@
+    "\\forall t: \\sum_{g} p_{\\text{reserve}}(g,t) \\geq \\text{reserve requirement}\n",
+    "$$\n",
+    "\n",
+    "- a constraint to ensure that the reserve power a generator provides must be less or equal than the difference between its output $p$ and its nominal capacity $p_\\text{nom}$, multiplied with a scalar coefficient $a$. This coefficient can have any value between 0 and 1 and represents the technical availability of a generator to provide balancing power. \n",
+    "$$\n",
+    "\\forall g, t: p_\\text{reserve}(g, t) \\leq a(g) p_\\text{nom}(g) - p(g,t)\n",
+    "$$\n",
+    "\n",
+    "- a constraint  to ensure that the balancing power a generator provides must be less or equal than its actual output $p$, multiplied with a scalar coefficient $b$. This coefficient can have any value between 0 and 1 and represents the technical availability of a generator to provide balancing power. \n",
+    "\n",
+    "$$\n",
+    "\\forall g, t: p_\\text{reserve}(g, t) \\leq b(g) p(g,t)\n",
+    "$$\n",
+    "\n",
+    "The relationships between the variables $a$, $b$, $p_\\text{nom}$, $p$, and $p_\\text{reserve}$ are depicted in the following schematic graph.\n",
+    "\n",
+    "![balancing-reserves-graph](../../doc/img/balancing-reserves-graph.png)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Limitations and other approaches\n",
+    "\n",
+    "Note that this is a highly simplified approach that has significant limitations:\n",
+    "- It does not distinguish between different categories of reserve power (primary, secondary and xx reserves). \n",
+    "- Reserves are provided symmetrical; there is no distinction between positive and negative reserves\n",
+    "- The approach only takes into account the provision of balancing power, but not the actual call for balancing power\n",
+    "\n",
+    "All these issues can be taken into account in a MIP unit commitment model, albeit at much higher numerical costs. \n",
+    "\n",
+    "Also note that reserve margin constraints have been added to PyPSA-EUR: https://github.com/PyPSA/pypsa-eur/blob/662492a23e4b0fe84f8d65611aad6668488aa88c/scripts/solve_network.py#L393-L472\n",
+    "\n",
+    "\n",
+    "## Implementation\n",
+    "We start by importing the required packages:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import matplotlib.pyplot as plt\n",
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "\n",
+    "import pypsa"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### The basic model\n",
+    "\n",
+    "Our toy model consits of a single bus with generators that have different marginal costs. We use a sine function for the load profile."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def setup_network():\n",

```

Comment:
I think for the example it works better if the model building is not encapsulated by a function.